### DISCUSSION TITLE
Testcontainers dotnet: store videos for failed tests onto local disc or upload to storage account

### DISCUSSION THREAD

**Discussion by ossentoo** [testcontainers/testcontainers-dotnet#787] [Q&A] [answered]
Hi,

I would like to store (or store and upload) the videos of failed tests to first local disk and then upload these to an azure storage account (this last bit I can handle).

Is there a way I can mount a volume into the video container (or some other method) to gain access to it?

**Comment by HofmeisterAn** [comment, marked as answer]
You can use `WithBindMount(string, string)` to mount a host path into the container. Please note this is considered a bad practice.

Read the file straight from the container with `ReadFileAsync(string, CancellationToken)` or download it from an HTTP endpoint.

**Reply by ossentoo** [reply]
thanks @HofmeisterAn .  Can you point me to some sample code somewhere that i can look at for either of the two examples above?  I'm struggling to understand where the source file would reside.

**Reply by HofmeisterAn** [reply]
Here is an [example](https://github.com/testcontainers/testcontainers-dotnet/blob/fe9f55a84341b57f5e054e631043f590dd2cc74c/tests/Testcontainers.Tests/Fixtures/Containers/Unix/ProtectDockerDaemonSocket.cs#L36) that mouts a host directory, and [here](https://github.com/testcontainers/testcontainers-dotnet/blob/fe9f55a84341b57f5e054e631043f590dd2cc74c/tests/Testcontainers.Tests/Unit/Containers/Unix/CopyResourceMappingContainerTest.cs#L56-L58) is one that reads a file from inside the container. It receives and array of bytes. If your file is large, we should probably support a stream too.

**Reply by ossentoo** [reply]
thanks a lot.

**Reply by ossentoo** [reply]
So, i've tried that - but it seems the video container is not creating video files.  I have code like this in my test constructor:

```
       _gridNetwork = new NetworkBuilder()
            .WithName($"gridNetwork{Guid.NewGuid().ToString()}")
            .Build();

        _containerHub = new ContainerBuilder()
            .WithImage("selenium/hub:latest")
            .WithName("selenium-hub")
            .WithPortBinding(4442, 4442)
            .WithPortBinding(4443, 4443)
            .WithPortBinding(SessionPort, SessionPort)
            .WithNetwork(_gridNetwork)
            .WithNetworkAliases(nameof(_containerHub))            
            .WithWaitStrategy(Wait.ForUnixContainer().UntilPortIsAvailable(SessionPort))         
            .Build();

        var firefoxEnvironment = new Dictionary<string, string>()
        {
            {"SE_EVENT_BUS_HOST", "selenium-hub"},
            {"SE_EVENT_BUS_PUBLISH_PORT", "4442"},
            {"SE_EVENT_BUS_SUBSCRIBE_PORT", "4443"}
        };

        _containerFirefox = new ContainerBuilder()
            .WithImage("selenium/node-firefox:latest")
            .WithEnvironment(firefoxEnvironment)
            .WithNetwork(_gridNetwork)  
            .WithNetworkAliases(nameof(_containerFirefox))            
            .Build();

        var firefoxVideoEnvironment = new Dictionary<string, string>()
        {
            {"DISPLAY_CONTAINER_NAME", "firefox"},
            {"FILE_NAME", VideoFileName}
        };      
           
        _containerVideo = new ContainerBuilder()
            .WithImage("selenium/video:latest")
            .WithNetwork(_gridNetwork)            
            .WithNetworkAliases(nameof(_containerVideo))     
            .WithBindMount("/tmp/videos", "/videos", AccessMode.ReadWrite)                   
            .WithEnvironment(firefoxVideoEnvironment)
            .Build(); 
```

then i initialize like this:

```
  public async Task InitializeAsync()
    {
        await _gridNetwork.CreateAsync().ConfigureAwait(false);
        await _containerHub.StartAsync().ConfigureAwait(false);
        await _containerFirefox.StartAsync().ConfigureAwait(false);
        await _containerVideo.StartAsync().ConfigureAwait(false);
        WebDriver = new RemoteWebDriver(new UriBuilder(Uri.UriSchemeHttp, _containerHub.Hostname, _containerHub.GetMappedPublicPort(SessionPort)).Uri, new FirefoxOptions());

    }
```

and then later on, i dispose of the containers like this:

```
public async Task DisposeAsync()
{    
        await _containerVideo.DisposeAsync().ConfigureAwait(false);
        await _containerFirefox.DisposeAsync().ConfigureAwait(false);
        await _containerHub.DisposeAsync().ConfigureAwait(false);
        await _gridNetwork.DeleteAsync().ConfigureAwait(false);
}
```

The tests do work as expected, and i can access VNC on the browser on port 4444 to see the video working.
After the disposal of the containers, however, when i check in the folder on the local machine, i can't see any video file.  Also, I've tried the container.ReadFileAsync() approach, but this doesn't work either (i get a file not found exception).  So it seems that the file isn't being created.  Does any one have any idea why and can you put me right please?

If there is a fully working sample with video saving, it would really help me to get this going. 

thanks

**Reply by HofmeisterAn** [reply]
Could you kindly format your question and code properly? It would be helpful for those who are trying to assist you. As someone who is seeking help, it would be appreciated if you could put in a minimum effort.

I assume min. your `DISPLAY_CONTAINER_NAME` is wrong. Your Firefox container is available under a different name. I think it should be: `{"DISPLAY_CONTAINER_NAME", nameof(_containerFirefox)}`. Please check you are using the right configuration values.

**Reply by ossentoo** [reply]
re-formatted.

**Reply by ossentoo** [reply]
thanks
